cli/lib/kontena/scripts/completer.rb

```diff
@@ -247,31 +279,40 @@ def subcommand_tree(cmd = nil, base = nil)
       when 'external-registry'
         completion.clear
         completion.push %w(add list delete)
-      when 'app'
-        completion.clear
-        sub_commands = %w(init build config deploy start stop remove rm ps list
-                          logs monitor show)
-        if words[1] && sub_commands.include?(words[1])
-          completion.push helper.yml_services
-        else
-          completion.push sub_commands
-        end
       when 'stack'
         completion.clear
-        sub_commands = %w(build install upgrade deploy start stop remove rm ls list
+        sub_commands = %w(build install upgrade deploy start stop remove restart list
                           logs monitor show registry inspect)
         if words[1]
-          if words[1] == 'registry'
-            registry_sub_commands = %(push pull search show rm)
-            completion.push registry_sub_commands
-          elsif %w(install).include?(words[1])
-              completion.push helper.yml_files
-          elsif words[1] == 'upgrade' && words[3]
+          if words[1] == 'registry' || words[1] == 'reg'
+            registry_sub_commands = %(push pull search show remove)
+            if words[2]
+              if words[2] == 'push'
+                completion.push helper.yml_files
+              elsif %w(pull search show remove rm).include?(words[2]) && words[4].nil?
+                completion.push helper.registry_stacks(words[3].to_s)
+              else
+                completion.push registry_sub_commands
+              end
+            else
+              completion.push registry_sub_commands
+            end
+          elsif %w(install validate build).include?(words[1])
             completion.push helper.yml_files
-          elsif words[1] && sub_commands.include?(words[1])
+            if words[1] == 'install'
+              completion.push helper.registry_stacks(words[2].to_s)
+            end
+          elsif words[1] == 'upgrade'
+            if words[3]
+              completion.push helper.yml_files
+              completion.push helper.registry_stacks(words[4].to_s)
+            else
+              completion.push.helper.stacks
```

Comment:
Extra `.` instead of ` `

```
DEBUG undefined method `helper' for []:Array (NoMethodError)
/home/kontena/kontena/kontena/cli/lib/kontena/scripts/completer.rb:310:in `<top (required)>'
bin/kontena:13:in `require'
bin/kontena:13:in `<top (required)>'
```